### DISCUSSION TITLE
Authorization for httpfs servers

### DISCUSSION THREAD

**Discussion by chrisbrain** [duckdb/duckdb#4963] [General]
Hi - Would it be possible to add in a mechanism to send say a HTTP Authorization header to httpfs servers in a similar way that keys/secrets can be set for s3?

For example you may wish to run a custom httpfs server to deliver parquet data and alter the data based on the user.

**Comment by hannes** [comment]
Sure, happy to review a PR

**Comment by Mause** [comment]
> Hi - Would it be possible to add in a mechanism to send say a HTTP Authorization header to httpfs servers in a similar way that keys/secrets can be set for s3?
> 
> For example you may wish to run a custom httpfs server to deliver parquet data and alter the data based on the user.

Is this not already possible, with basic auth in the URL itself?

**Comment by chrisbrain** [comment]
Doesn't seem to work no @Mause - I think that technique might have fallen out of favour for security reasons or something.

At least when I try:
`select * from 'http://user:pwd@localhost:8888/mytestserver/data.parquet';`

I am getting:
`Error: Invalid Error: HTTP HEAD error on ... (Error code 2)`

**Reply by Mause** [reply]
True, it looks like our http client library doesn't support it at all. Shouldn't be too difficult to add support however, would be similar to my [http proxy pr](https://github.com/duckdb/duckdb/pull/4741/files)

**Comment by tamargrey** [comment]
@Mause @hannes I'm looking to be able to pass in http headers (not necessarily for authentication). Has there been anything added for this since the last comment above?

**Reply by Mause** [reply]
You can implement something with the fsspec support in the Python bindings, but the general case in DuckDB core hasn't been implemented no

**Reply by tamargrey** [reply]
Got it--I'd been trying to get fsspec to work but had been failing there as well. I put together something along the lines of

```python
import fsspec
import duckdb

endpoint = "http://localhost:3000/readFile?uri=file:///location/data.parquet"
headers = {"key": "value"}
fs = fsspec.filesystem("http", headers=headers)
duckdb.register_filesystem(fs)

create_table_query = f"SELECT * from '{endpoint}'"
result  = duckdb.sql(create_table_query)
```
That gets me a `FileNotFoundError`, which I'm hoping is just user error on my part. Does this stand out to you as having anything wrong in it?

I know that I can access that data just fine with pandas' `read_parquet` util, so I'm confident that the file does exist (I need DuckDB specifically so I can query for a limited number of rows, which `read_parquet` doesn't support).

**Reply by Mause** [reply]
Perhaps the typo in the host name?

**Reply by Tishj** [reply]
> localholst

**Reply by tamargrey** [reply]
Oops! That's just my own typo from retyping this out instead of copy/pasting my actual endpoint. Edited to fix.

**Reply by Tishj** [reply]
There's a chance our HTTPFS extension is interfering with this, as `http` is also a host that it looks for to forward the call to its filesystem

**Reply by tamargrey** [reply]
Interesting. So it may be impossible to use an `http` fsspec filesystem with duckdb? 

Also, as a sanity check: That typo would produce the error `duckdb.IOException: IO Error: No files found that match the pattern`. And I can see on my server's logs that the requests are being made successfully and that the header is doing what I need it to. It's just the matching of the file. I'm wondering if there's some other way I need to be formatting the request string?

**Reply by Tishj** [reply]
I remember an issue someone made about this, but in general people won't have their own http filesystem so it was disregarded at the time.
It's possible to assign the fsspec a different host, though it might not be trivial

Have a look at what we did for our internal memory filesystem in the python client:
`tools/pythonpkg/pyduckdb/filesystem.py`
https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/pyduckdb/filesystem.py

**Reply by Mause** [reply]
```py
import fsspec                                                         
import duckdb                                           
endpoint = "http://api.ipify.org/?format=json"          
headers = {"key": "value"}
fs = fsspec.filesystem("http", headers=headers)         
duckdb.register_filesystem(fs)
create_table_query = f"SELECT * from read_json_auto('{endpoint}')"
result  = duckdb.sql(create_table_query)
```

This gets you 90% there, the last part being the seekable bug fix that was just merged: https://github.com/duckdb/duckdb/pull/6585

**Reply by tamargrey** [reply]
@Mause Nice! I see that that snippet with `fsspec` 2023.5.0 and `duckdb` 0.7.1 produces a corresponding `FileNotFoundError`. I'm not set up right now to confirm that this fixes my use case, but let me know if that'd be useful and I can try to install these changes locally. Do y'all have a timeline for releasing the fix?

**Reply by Mause** [reply]
@tamargrey hopefully in the next week or so, with the next bugfix release! It should be in tomorrow's nightly/dev release at the very least
